### ISSUE TITLE
certonly does not run renewal-hooks/deploy scripts

### ISSUE THREAD

**Issue by SystemParadox** [certbot/certbot#5935]
Hi.

I got certbot working with a renewal hook script in /etc/letsencrypt/renewal-hooks/deploy/haproxy.sh, which runs correctly with `certbot renew`.

However, I was surprised to find it did not run my hook when I added a new certificate with `certbot certonly`.

I would like to add that I have found the documentation regarding hooks to be very confusing and difficult to find. It would be worth adding a hooks section to the documentation explaining exactly which hooks are run when and which if any are considered deprecated.

I am running Ubuntu 14.04, certbot 0.22.2 (installed via PPA).

Thanks.

**Comment by bmw** [comment]
Sorry for my very delayed reply @SystemParadox. It's been a busy couple of weeks.

The behavior you're seeing where hooks in `/etc/letsencrypt/renewal-hooks` only run with the `renew` subcommand is intentional. If you want the hooks to run with other subcommands such as `certonly` or `run`, I'd recommend removing them from the `renewal-hooks` directory and specifying them in https://certbot.eff.org/docs/using.html#configuration-file.

I agree that the documentation for this needs some work. This behavior is more or less documented at https://certbot.eff.org/docs/using.html#renewing-certificates, however, I agree it's not very clear. I'd like to hijack this issue to track improving this documentation, but if you disagree please let me know.

To answer your question about deprecation, only the `--renew-hook` flag is deprecated.

**Comment by SystemParadox** [comment]
Thanks for the reply. If this is intentional it would make sense to turn this into an 'update documentation' issue.

**Comment by berk-can** [comment]
I also wonder what permissions and owner of hook scripts should be

**Comment by nspassov** [comment]
Apologies for resurrecting the dead, but I have related questions.

Do I understand correctly that pre/post/deploy hooks can be specified in the renewal config? Would the correct key names in the .conf file be `pre_hook`/`post_hook`/`deploy_hook`? Could the value be either a path to a script as well as a shell command?

Do I understand correctly that the deploy hook (or any of the hooks?) would not be run when dry-run option is used?

**Comment by bmw** [comment]
> Do I understand correctly that pre/post/deploy hooks can be specified in the renewal config?

Yes, however, I recommend letting Certbot preserve these options for you. These files weren't really designed to be edited by the user.

If you obtain a certificate and have pre/post/deploy hooks specified on the command line or in a a [global configuration file](https://certbot.eff.org/docs/using.html#configuration-file), those values will be automatically written to the renewal configuration file under `/etc/letsencrypt/renewal` to be reused when you renew the certificate without these flags.

You can also specify hooks by placing executables under `/etc/letsencrypt/` as described at https://certbot.eff.org/docs/using.html#renewing-certificates. The relevant paragraph there is:

> You can also specify hooks by placing files in subdirectories of Certbot’s configuration directory. Assuming your configuration directory is /etc/letsencrypt, any executable files found in /etc/letsencrypt/renewal-hooks/pre, /etc/letsencrypt/renewal-hooks/deploy, and /etc/letsencrypt/renewal-hooks/post will be run as pre, deploy, and post hooks respectively when any certificate is renewed with the renew subcommand. These hooks are run in alphabetical order and are not run for other subcommands. (The order the hooks are run is determined by the byte value of the characters in their filenames and is not dependent on your locale.)

> Could the value be either a path to a script as well as a shell command?

Yes. The value given to these flags can be either a path or a shell command.

> Do I understand correctly that the deploy hook (or any of the hooks?) would not be run when dry-run option is used?

Pre and post hooks will be run with `--dry-run`, however, the deploy hook will not be.

**Comment by npodbielski** [comment]
Maybe this is not ideal place but I seem to unable to find information how those hook works. 
What I am trying to achieve is to copy cert files to some temp dir readable to my user instead of root. Sending them via ssh would be fine too but I am not sure what user is used to run certbot (seems like it is root right?). Anyway script with copy/sending seems to be not running at all. I tried to use `deply-hook` and `post-hook`. I did not try to use hook directory because I am interested only in one domain hook not all of them.
My certificates are generated with ovh plugin and installed on nginx. Maybe those plugins use hooks and mine are ignored? Any ideas?

**Comment by bmw** [comment]
The best documentation we have right now is in this issue and its links to various bits of our documentation.

If you need further help, I recommend posting to https://community.letsencrypt.org where there is a large community of people familiar with Certbot who should be able to help you get things set up.

**Comment by npodbielski** [comment]
Thx. Will do.

**Comment by iam-TJ** [comment]
For those who find this issue whilst trying to resolve renewal config file problems, here's something we've learned today.

certbot 0.40 (Ubuntu 20.04)

We have an /etc/letsencrypt/renewals/domain.tld.conf

which contained some options with underscores in their name but we were unable to find any documentation that underscores ("_") rather than hyphens ("-") should be used in the .conf file (man certbot and online documents hyphenated options only).

We also found an auto-generated "renew_hook =" (created when the domain was first added to certbot) which we have been unable to find documentation on. We needed to replace that with pre and post hooks to stop and start a TURN/STUN service using ports 80 and 443 to allow  certbot standalone mode.

According to the docs we added "pre-hook =" and "post-hook =" which failed; we guessed correctly they required underscores "pre_hook =" and "post_hook =".

**Comment by kbkaiser** [comment]
FWIW (running 0.31), I issued a command  certbot renew --force-renewal --deploy-hook "service dovecot reload"

This caused   

renew_hook = service dovecot reload  

(not deploy_hook) lines to be added to the renewals .conf files. N.B: no quotes, spacing as shown

**Comment by jjakob** [comment]
I agree, since I couldn't find the documentation for certificate renewal config files (in `/etc/letsencrypt/renewal/$domain.conf`) I looked through the source, if you specify `--deploy-hook` on the command line it is stored in the conf file as `renew_hook`.  Since 2.3.0 there's no need to edit the config files directly since there is a "certbot reconfigure" subcommand which can change the deploy hook, but if you happen to have an older version or are writing a script that should support older versions, manual config file editing is still necessary.
https://github.com/certbot/certbot/blob/23090198bf132704f1c860d5b6b623a2fcb593f8/certbot/certbot/_internal/cli/cli_constants.py#L83
https://github.com/certbot/certbot/blob/a3c9371dc5d3f51b1382da84acc3bb42bea8426b/certbot/tests/main_test.py#L633-L634

**Comment by npodbielski** [comment]
It is good for me. I was able to make it work with hooks.

**Comment by fabriziocom** [comment]
> Hi.
> 
> I got certbot working with a renewal hook script in /etc/letsencrypt/renewal-hooks/deploy/haproxy.sh, which runs correctly with `certbot renew`.
> 
> However, I was surprised to find it did not run my hook when I added a new certificate with `certbot certonly`.
> 
> I would like to add that I have found the documentation regarding hooks to be very confusing and difficult to find. It would be worth adding a hooks section to the documentation explaining exactly which hooks are run when and which if any are considered deprecated.
> 
> I am running Ubuntu 14.04, certbot 0.22.2 (installed via PPA).
> 
> Thanks.

e se
